solarforecastarbiter/reference_forecasts/main.py

```diff
@@ -465,3 +465,143 @@ def make_latest_nwp_forecasts(token, run_time, issue_buffer, base_url=None):
+    forecast must have a is_reference_persistence_forecast=true and an
+    observation_id in Forecast.extra_parameters (formatted as a JSON
+    string). A boolean value for "index_persistence" in
+    Forecast.extra_parameters controls whether the persistence
+    forecast should be made adjusting for clear-sky/AC power index or
+    not.
+
+    Parameters
+    ----------
+    session : solarforecastarbiter.io.api.APISession
+    forecasts : list of datamodel.Forecasts
+        The forecasts that should be filtered to find references.
+    observations : list of datamodel.Observations
+        Observations that will are available to use to fetch values
+        and make persistence forecasts.
+    max_run_time : pandas.Timestamp
+        The maximum run time/issue time for any forecasts. Usually now.
+
+    Returns
+    -------
+    generator of (Forecast, Observation, next_issue_time, index)
+
+    """
+    user_info = session.get_user_info()
+    observation_dict = {obs.observation_id: obs for obs in observations}
+    for fx in forecasts:
+        if not _is_reference_persistence_forecast(fx.extra_parameters):
+            logger.debug(
+                'Forecast %s is not labeled as a reference '
+                'persistence forecast',  fx.forecast_id)
+            continue
+
+        if not fx.provider == user_info['organization']:
+            logger.debug(
+                "Forecast %s is not in user's organization",
+                fx.forecast_id)
+            continue
+
+        try:
+            extra_parameters = json.loads(fx.extra_parameters)
+        except json.JSONDecodeError:
+            logger.warning(
+                'Failed to decode extra_parameters for %s: %s as JSON',
+                fx.name, fx.forecast_id)
+            continue
+
+        try:
+            observation_id = extra_parameters['observation_id']
+        except KeyError:
+            logger.error(
+                'Forecast, %s: %s, has no observation_id to base forecasts'
+                ' off of. Cannot make forecast.',
+                fx.name, fx.forecast_id)
+            continue
+        if observation_id not in observation_dict:
+            logger.error(
+                'Observation %s not in set of given observations.'
+                ' Cannot generate persistence forecast for %s:%s.',
+                observation_id, fx.name, fx.forecast_id)
+            continue
+        observation = observation_dict[observation_id]
+
+        index = extra_parameters.get('index_persistence', False)
+        obs_mint, obs_maxt = session.get_observation_time_range(observation_id)
+        if pd.isna(obs_maxt):  # no observations to use anyway
+            logger.info(
+                'No observation values to use for %s:%s from observation %s',
+                fx.name, fx.forecast_id, observation_id)
+            continue
+
+        fx_mint, fx_maxt = session.get_forecast_time_range(fx.forecast_id)
```

Comment:
what happens if there's a time span in which forecasts are not created because of something like delayed data ingest (e.g. the SURFRAD problems) or a crashed worker and then they start again? Is it possible to fill that gap?

Reply:
that's actually the whole point of looking at ``fx_maxt``. We find the next issue time based on it to go from ``fx_maxt`` to ``fx_maxt + run_length`` and then continue up to some max time. In the while loop, we also make sure all the data required for a forecast is available before making it.

For data gaps, we don't go back and make the persistence forecast if the data is filled in, but I think that should just be dealt with separately

Reply:
My only concern is that we have the ability to easily deal with gaps once they are created and I'm wondering if this function should be modified to support that.

Reply:
I think that is a separate issue and interface. This already handles the case where data is automatically fetched although there may be a delay or gaps. To fill in the data gaps, we already need to manually run a cli script, so it would be reasonable to fill in the persistence fx gaps in the same way.

We also don't want this to try and automatically fill in the forecast gaps everytime it runs. There was likely a reason the gaps exist and we would likely hit the same error over and over.

Reply:
well adding in a cli function to run this and possible specifying ``--start`` like ``referencenwp`` and ``referencedata`` wouldn't pollute this PR too much. But we should also be able fetch the forecast, find missing periods, and make those missing series from a CLI script, which I think belongs in a new pr

Reply:
Ok that makes sense.

Reply:
https://github.com/SolarArbiter/solarforecastarbiter-core/issues/459